Updated dockerfile, switched dependency management from pip to poetry

HughieH opened the pull request:
^ details in title, but additionally:

- deleted old dockerfile
- using ubuntu 22.04 instead of 20.04
- installed GDAL as a system dependency

HughieH committed:
poetry and dockerfile updates

HughieH committed:
deleted dockerfile

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,70 @@
+docopt = "0.6.2"
+filelock = "3.8.0"
+Flask = "2.2.5"
+Flask-Cors = "3.0.10"
+Flask-Sock = "*"
+idna = "3.4"
+importlib-metadata = "5.0.0"
+itsdangerous = "2.1.2"
+Jinja2 = "3.1.2"
+joblib = "1.2.0"
+MarkupSafe = "2.1.1"
+marshmallow = "3.18.0"
+mercantile = "1.2.1"
+numpy = "1.23.4"
+packaging = "21.3"
+pandas = "1.5.1"
+Pillow = "10.0.1"
+pipenv = "2022.10.25"
+pipreqs = "0.4.13"
+platformdirs = "2.5.2"
+plotly = "5.11.0"
+pyparsing = "3.0.9"
+python-dateutil = "2.8.2"
+pytz = "2022.6"
+PyYAML = "6.0"
+rasterio = "1.3.3"
+requests = "2.31.0"
+Shapely = "1.8.5.post1"
+six = "1.16.0"
+snuggs = "1.4.7"
+tenacity = "8.1.0"
+terracotta = "0.7.5"
+toml = "0.10.2"
+tqdm = "4.64.1"
+urllib3 = "1.26.14"
+virtualenv = "20.16.6"
+virtualenv-clone = "0.5.7"
+Werkzeug = "2.2.3"
+yarg = "0.1.9"
+zipp = "3.10.0"
```
Does this accurately reflect the true top-level requirements of this package?  This smells like a `pip freeze` dump!

HughieH replied:
Most of these dependencies aren't being used right now, but Jason mentioned that they will be required later on when he onboards code from the ML side onto the classification service. 

Is it okay if we leave this as is temporarily? Currently server.py only really uses flask.

ntlhui replied:
It would be better to include those dependencies at that time, not now.

ntlhui replied:
Leaving these in may cause very difficult to resolve conflicts later on!  @Jasonnyang @zclawr please make note of this before moving forward.  This may also cause a lot of difficulty in building this outside of `docker`!

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,70 @@
+[tool.poetry]
+name = "MM-classification-service"
+version = "0.1.0"
+description = ""
+authors = []
```
You seem to be missing package information

HughieH replied:
As mentioned before, I plan on updating package and dependancy information once Jason onboards the ML side of our classification service.

ntlhui commented on pyproject.toml:
You seem to be missing entry point information

ntlhui replied:
You seem to be missing a lock file!

HughieH replied:
added lock file and entrypoint information.

ntlhui commented on Dockerfile:
```diff
@@ -1,23 +1,32 @@
-# Copy files to working directory
-COPY . .
+# Copy only pyproject.toml and poetry.lock (if exists) to cache dependencies
+COPY pyproject.toml poetry.lock* /app/
```
`poetry.lock` must always exist!

HughieH replied:
Generated poetry.lock file based on current dependencies.

ntlhui replied:
`poetry.lock` should be committed to git

ntlhui commented on Dockerfile:
```diff
@@ -1,23 +1,32 @@
+RUN if [ $(gdal-config --version | cut -d '.' -f1,2) \< "3.1" ]; then echo $(gdal-config --version) && exit 1; fi
+
+# Install Poetry
+RUN pip3 install poetry
```
Do not invoke `pip3` like this - always invoke as `python3.x -m pip` to ensure you are using the correct python!

HughieH replied:
Alright will do. This image uses Python 3.10

ntlhui commented on Dockerfile:
```diff
@@ -1,23 +1,32 @@
 # Start the application
-CMD ["python3", "server.py"]
\ No newline at end of file
+CMD ["python3", "server.py"]
```
Use a poetry provided entry point!

HughieH replied:
added a poetry entrypoint

HughieH committed:
Resolving comments from Nathan

ntlhui commented on pyproject.toml:
```diff
@@ -1,8 +1,13 @@
+
+# poetry entrypoint for running our Flask server
+[tool.poetry.scripts]
+classification-server = "server:main"
```
`server.py` has no `main` function!

ntlhui commented:
I am currently unable to build and install this package!
```
python3.8 -m venv .venv
source .venv\bin\activate
python -m pip install --upgrade pip poetry
poetry install
```
This results in a failure to install `terracotta`.  This doesn't seem to be imported by any python files or executed by the Dockerfile!

Removing `terracotta` using `poetry remove terracotta` then running `poetry install` results in `No file/folder found for package mm-classification-service`.

If the intent is to use `poetry` to do dependency and package management, then this project must be structured as a python package.

See https://github.com/UCSD-E4E/python-repo-example for how this repo should be structured.

HughieH commented:
Alright, I'll restructure our classification service as a package instead of just directly running an executable. I was running into issues with creating a poetry entrypoint without structuring my code as a package but doing this should resolve my issues.

HughieH committed:
configured toml file and converted service into being a python package

HughieH commented:
## Updates:
- Updated file structure into being a python package as specified in the E4E python package repo example.
- Added poetry entrypoint and some configurations in the Dockerfile to allow us to use this entrypoint.
- Added some changes to the Dockerfile including adding a symlink for python3 and adjusting how our layers are built.

HughieH committed:
Updates to readme

ntlhui commented on Dockerfile:
```diff
@@ -1,23 +1,37 @@
-CMD ["python3", "server.py"]
\ No newline at end of file
+# Run our Flask app using poetry entrypoint
+ENTRYPOINT ["poetry", "run", "classification-server"]
```
You should see if you can execute without using `poetry run` - by installing using `poetry`, this should create a `classification-server` executable that you can invoke.

HughieH replied:
changed entrypoint to this

ntlhui commented on mm_classification/server.py:
```diff
@@ -113,8 +113,10 @@ def classify(array, classifier_id):
-if __name__ == '__main__':
+# callable function to run server, included this in order to define our poetry entrypoint
+def main():
     app.run(debug=False, port=5001, host='0.0.0.0')
```
Is this necessarily correct?  When invoking, this indicates running on several interfaces!

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,75 @@
+[tool.poetry]
+name = "mm_classification"
+version = "0.1.0"
```
There are two sources of truth for this!  Other is `mm_classification/__init__.py`.  This is not ideal!

HughieH replied:
changed tool.petry.name to not be slightly different

ntlhui replied:
There are still two sources of truth for `tool.poetry.version`!

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,75 @@
+[tool.poetry]
+name = "mm_classification"
+version = "0.1.0"
+packages = []
```
This should probably explicitly list `mm_classification`

HughieH replied:
added this

ntlhui commented on pyproject.toml:
```diff
@@ -0,0 +1,75 @@
+version = "0.1.0"
+packages = []
+description = "This is the Mangrove Monitoring classification service. "
+authors = ["E4E - Mangrove Monitoring SWE Team"]
```
List out authors individually.

HughieH replied:
added this

HughieH committed:
Classification PR fix-ups

HughieH committed:
removed pacaking module
